src/Bicep.Decompiler/TemplateConverter.cs

```diff
@@ -0,0 +1,907 @@
+            {
+                var stringVal = flattenedJTokenExpression.Value.Value<string>();
+                if (stringVal == null)
+                {
+                    return null;
+                }
+
+                return SyntaxHelpers.CreateStringLiteral(stringVal);
+            }
+
+            if (flattenedExpression is not FunctionExpression functionExpression)
+            {
+                throw new InvalidOperationException($"Expected {nameof(FunctionExpression)}");
+            }
+            expression = functionExpression;
+
+            var values = new List<string>();
+            var stringTokens = new List<Token>();
+            var expressions = new List<SyntaxBase>();
+            for (var i = 0; i < functionExpression.Parameters.Length; i++)
+            {
+                // FlattenStringOperations will have already simplified the concat statement to the point where we know there won't be two string literals side-by-side.
+                // We can use that knowledge to simplify this logic.
+
+                var isStart = (i == 0);
+                var isEnd = (i == functionExpression.Parameters.Length - 1);
+
+                if (functionExpression.Parameters[i] is JTokenExpression jTokenExpression)
+                {
+                    stringTokens.Add(SyntaxHelpers.CreateStringInterpolationToken(isStart, isEnd, jTokenExpression.Value.ToString()));
+
+                    // if done, exit early
+                    if (isEnd)
+                    {
+                        break;
+                    }
+                    // otherwise, process the expression in this iteration
+                    i++;
+                }
+                else
+                {
+                    //  we always need a token between expressions, even if it's empty
+                    stringTokens.Add(SyntaxHelpers.CreateStringInterpolationToken(isStart, false, ""));
+                }
+
+                expressions.Add(ParseLanguageExpression(functionExpression.Parameters[i]));
+                isStart = (i == 0);
+                isEnd = (i == functionExpression.Parameters.Length - 1);
+
+                if (isEnd)
+                {
+                    // always make sure we end with a string token
+                    stringTokens.Add(SyntaxHelpers.CreateStringInterpolationToken(isStart, isEnd, ""));
+                }
+            }
+
+            var rawSegments = Lexer.TryGetRawStringSegments(stringTokens);
+            if (rawSegments == null)
+            {
+                return null;
+            }
+            
+            return new StringSyntax(stringTokens, expressions, rawSegments);
+        }
+
+        private SyntaxBase ParseFunctionExpression(FunctionExpression expression)
+        {
+            SyntaxBase? baseSyntax = null;
+            switch (expression.Function.ToLowerInvariant())
+            {
+                case "parameters":
```

Comment:
How do we handle situations when there's parameter and variable with the same name? (Output could be a problem as well because right now we enforce uniqueness across all identifiers.)

Reply:
The nameResolver class handles that (note that we lookup the name on line 324 here using the nameResolver)